src/System.Reflection.Metadata/src/System/Reflection/Metadata/Handles.cs

```diff
@@ -2713,61 +2713,66 @@ public override int GetHashCode()
     /// </summary>
     public struct NamespaceDefinitionHandle : IEquatable<NamespaceDefinitionHandle>
     {
-        // At this time, IsVirtual is always false because namespace names come from type definitions 
-        // and type forwarders only, which never get their namespaces projected.
-        // 
-        // For Namespace, the offset is to the null-terminated full name of the 
-        // namespace in the #String heap.
+        // Non-virtual (namespace having at least one type or forwarder of its own) 
+        // heap offset is to the null-terminated full name of the namespace in the 
+        // #String heap.
         //
-        // For SyntheticNamespace, the offset points to the dot-terminated simple name of the namespace
-        // in the #String heap. This is used to represent namespaces that are parents of other namespaces
-        // but no type definitions or forwarders of their own.
+        // Virtual (namespace having child namespaces but no types of its own) 
+        // the virtual index is an auto-incremented value and serves solely to 
+        // create unique values for indexing into the NamespaceCache.
 
         // bits:
-        //     31: IsVirtual (0)
-        // 29..31: type (non-virtual: Namespace, SyntheticNamespace)
+        //     31: IsVirtual
+        // 29..31: 0
         //  0..28: Heap offset or Virtual index
         private readonly uint _value;
 
         private NamespaceDefinitionHandle(uint value)
         {
-            // note: no virtual namespaces allowed
-            Debug.Assert((value & HeapHandleType.TypeMask) == NamespaceHandleType.Namespace ||
-                         (value & HeapHandleType.TypeMask) == NamespaceHandleType.SyntheticNamespace);
-
             _value = value;
         }
 
         internal static NamespaceDefinitionHandle FromFullNameOffset(int stringHeapOffset)
         {
-            return new NamespaceDefinitionHandle(NamespaceHandleType.Namespace | (uint)stringHeapOffset);
+            return new NamespaceDefinitionHandle((uint)stringHeapOffset);
         }
 
-        internal static NamespaceDefinitionHandle FromSimpleNameOffset(int stringHeapOffset)
+        internal static NamespaceDefinitionHandle FromVirtualIndex(uint virtualIndex)
         {
-            return new NamespaceDefinitionHandle(NamespaceHandleType.SyntheticNamespace | (uint)stringHeapOffset);
+            // we arbitrarily disallow 0 virtual index to simplify nil check.
+            Debug.Assert(virtualIndex != 0); 
+
+            if (!HeapHandleType.IsValidHeapOffset(virtualIndex))
+            {
+                // only a pathological assembly would hit this, but it must fit in 29 bits.
+                ThrowTooManySubnamespaces();
+            }
+
+            return new NamespaceDefinitionHandle(TokenTypeIds.VirtualBit | virtualIndex);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
```

Comment:
Yes. Either that or FromVirtualIndex gets inlined but uses unnecessary space in its caller. A similar pattern applied to BlobReader made a very measurable difference. With that said, this case is not very highly travelled and it probably doesn't matter here.  I find it easier to keep the whole library consistent than to decide on each throw on a case-by-case basis. 

I've filed #2261 to re-organize so that the approach is centralized and documented.